File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& prerequisites) {
        // topological sort
        vector<vector<int>> AL(n);
        vector<int> indegree(n,0);
        vector<int> ans;

        for (auto x : prerequisites) {
            AL[x[1]].push_back(x[0]);
            indegree[x[0]]++;
        }

        queue<int> q;
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            auto node = q.front();
            ans.push_back(node);
            q.pop();

            for (auto neighbour : AL[node]) {
                indegree[neighbour]--;
                if (indegree[neighbour] == 0) {
                    q.push(neighbour);
                }
            }
        }

        return ans.size() == n;
    }
};

File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # topo sort using kahn's

        AL = defaultdict(list) # store prereq -> course
        inDegree = defaultdict(int) # store count
        
        for course, prereq in prerequisites:
            AL[prereq].append(course)
            inDegree[course] += 1
        
        q = deque()

        for i in range(numCourses):
            if inDegree[i] == 0:
                q.append(i)
        
        visited = 0
        while q:
            curr = q.popleft()
            visited += 1
            for neighbour in AL[curr]:
                inDegree[neighbour] -= 1
                if inDegree[neighbour] == 0:
                    q.append(neighbour)
        
        return visited == numCourses